src/test_fixture/transport/channeled_transport.rs

```diff
@@ -0,0 +1,670 @@
+
+use crate::{
+    helpers::{
+        query::QueryConfig,
+        transport::{
+            ChannelledTransport, NoResourceIdentifier, QueryIdBinding, RouteId, RouteParams,
+            StepBinding,
+        },
+        HelperIdentity,
+    },
+    protocol::{QueryId, Step},
+};
+use async_trait::async_trait;
+use futures::{Stream, StreamExt};
+use futures_util::stream;
+use serde::de::DeserializeOwned;
+use std::{
+    collections::{hash_map::Entry, HashMap, HashSet},
+    fmt::{Debug, Formatter},
+    future::Future,
+    io,
+    pin::Pin,
+    sync::{Arc, Mutex},
+    task::{Context, Poll, Waker},
+};
+use tokio::sync::mpsc::{channel, Receiver, Sender};
+use tokio_stream::wrappers::ReceiverStream;
+use tracing::Instrument;
+
+type ConnectionTx = Sender<(Addr, InMemoryStream)>;
+type ConnectionRx = Receiver<(Addr, InMemoryStream)>;
+type StreamItem = Vec<u8>;
+
+/// In-memory implementation of [`ChannelledTransport`] backed by Tokio mpsc channels.
+/// Use [`Setup`] to initialize it and call [`Setup::start`] to make it actively listen for
+/// incoming messages.
+struct InMemoryChannelledTransport {
+    identity: HelperIdentity,
+    connections: HashMap<HelperIdentity, ConnectionTx>,
+    record_streams: StreamCollection<InMemoryStream>,
+}
+
+impl InMemoryChannelledTransport {
+    fn new(identity: HelperIdentity, connections: HashMap<HelperIdentity, ConnectionTx>) -> Self {
+        Self {
+            identity,
+            connections,
+            record_streams: StreamCollection::default(),
+        }
+    }
+
+    /// TODO: maybe it shouldn't be active, but rather expose a method that takes the next message
+    /// out and processes it, the same way as query processor does. That will allow all tasks to be
+    /// created in one place (driver). It does not affect the [`ChannelledTransport`] interface,
+    /// so I'll leave it as is for now.
+    fn listen<CB: ReceiveQueryCallback>(
+        &self,
+        mut callbacks: TransportCallbacks<CB>,
+        mut rx: ConnectionRx,
+    ) {
+        tokio::spawn(
+            {
+                let streams = self.record_streams.clone();
+                async move {
+                    let mut active_queries = HashSet::new();
+                    while let Some((addr, stream)) = rx.recv().await {
+                        tracing::trace!("received new message: {addr:?}");
+
+                        match addr.route {
+                            RouteId::ReceiveQuery => {
+                                let qc = addr.into::<QueryConfig>();
+                                let query_id = (callbacks.receive_query)(qc)
+                                    .await
+                                    .expect("Should be able to receive a new query request");
+                                assert!(
+                                    active_queries.insert(query_id),
+                                    "the same query id {query_id:?} is generated twice"
```

Comment:
Nice.  This way we can wire most tests with a fixed query ID generator, like https://xkcd.com/221/